### PR TITLE
config/perf: use prefix matchers instead of regexes

### PR DESCRIPTION
**Pull request by Augustyniak**
Commit Message: Converting regex matches to prefix matches as the former are way more costly. This is supposed to help with Envoy Mobile initialization times. The changes increase the number of stats that are enabled with default configuration of Envoy Mobile from 317 to 461. The profiling did not show any significant performance impact of this change but in theory reducing the number of places where we use regex should help.
Additional Description: related PR https://github.com/envoyproxy/envoy/pull/24521
Risk Level: Low
Testing: Manual
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

Signed-off-by: Rafal Augustyniak <raugustyniak@lyft.com>

### CODE REVIEW

**Comment by Augustyniak** [comment]
@jmarantz I was able to convert all but one matcher to `prefix`. The last one is tricky - I could try to convert is using `suffix` instead but I wanted to confirm whether using `suffix` has the same performance benefits as using `prefix`?

**Comment by jmarantz** [comment]
It's always better to use suffix-matching than a regex, but it doesn't have the same special-case code currently as `"prefix: foo."` (the trailing dot triggers the optimization). It would not be hard to add an optimization for `"suffix .foo" (leading dot would trigger).

The cases like `cluster.base.upstream_cx_` are better than using regex but will still be expensive, as the elaborated string representation of the stat names will need to be created, and the rejected names remembered in a map.

**Comment by jmarantz** [comment]
This new simplification looks great -- it errs on the side of disabling a wider subset of stats; assuming that's OK.

I think this could motivate someone -- perhaps in E-M , to optimize for the remaining safe-regex by only applying elaborating the string for stats whose StatName starts with the "vhost" token. That will greatly reduce the amount of string elaboration required.

But it would make sense to take some more flame-graphs with this PR in place first; this might be enough.

**Comment by jmarantz** [comment]
Another option is to punt on that vhost exclusion and just let that stat live :)

**Comment by Augustyniak** [comment]
> Another option is to punt on that vhost exclusion and just let that stat live :)

I definitely thought about doing this but in our case - Lyft - we have quite a few of virtual clusters and enabling all `vhost.*` stats leads to a significant increase in the total number of stats that need to be created (we are talking about thousands of stats in here). In general, we should make this inclusion list configurable (https://github.com/envoyproxy/envoy/issues/24459) so that each company can tweak it as needed and decide what trade-offs work best for them.

**Comment by Augustyniak** [comment]
>This new simplification looks great -- it errs on the side of disabling a wider subset of stats; assuming that's OK.

The PR increase the number of stats enabled by default in EM from 317 to 462 but I think that that increase is fine given the fact that we use regexes in only one place now.

**Comment by jmarantz** [comment]
The existence of *any* rejection expression that is not of the form `prefix: foo.` causes *all* stats to be name-elaborated so they can run through the string-matcher API.

How many more stats are added if you drop the vhost regex?

**Comment by Augustyniak** [comment]
>How many more stats are added if you drop the vhost regex?

For the Envoy Mobile configuration that I am interested in the numbers look as follows:
regex vhost matcher: 2336 stats
prefix vhost matcher: 6067 stats

**Comment by Augustyniak** [comment]
@jmarantz if you do not mind, I would like to merge this PR without further changes to `vhost` and collect some prod data to see whether the changes that I made in here helped or not. That being said I am a curious what you think about the `vhost` stats after learning about the increase in the number of stats I observed in https://github.com/envoyproxy/envoy/pull/24733#issuecomment-1370170222.

Merging this PR today as opposed to tmrw will speed up the process of collecting the data by a week since we are doing a beta/release cut later today.

**Review by alyssawilk — APPROVED** [review]

**Review by jmarantz — APPROVED** [review]

**Comment by jmarantz** [comment]
Checking this in now and iterating makes sense to me. This is a great improvement.

checkAndRememberRejection would drop out of this flame-graph if our rejections were just based on prefix tokens w/o regexes, but it's probably a lot smaller than it used to be due to this change.

File: mobile/library/common/config/config.cc [hunk 1]
```diff
@@ -496,28 +496,14 @@ stats_sinks: *stats_sinks
...
-        - safe_regex:
-            regex: '^cluster\.[\w]+?\.upstream_cx_[\w]+'
-        - safe_regex:
-            regex: '^cluster\.[\w]+?\.upstream_rq_[\w]+'
-        - safe_regex:
-            regex: '^cluster\.[\w]+?\.update_(attempt|success|failure)'
-        - safe_regex:
-            regex: '^cluster\.[\w]+?\.http2.keepalive_timeout'
-        - safe_regex:
-            regex: '^dns.apple.*'
```
Comment (Augustyniak) [inline mobile/library/common/config/config.cc:508]: disable dns.apple - I added it but it has never been used by anybody I think so should be safe to remove it.

**Pull request merged**